### ISSUE TITLE
Validator test case 3.5.15

### ISSUE THREAD

**Issue by tmarkey64** [DMTF/SPDM-Responder-Validator#138]
The Validator will flag an error for 3.5.15 if the capabilities MUT_AUTH_CAP is not set and baseAsymAlg is not zero, but the baseAsymAlg is set to the signature algorithm for the purpose of signature generation and verification.  There is not requirement in the specification that this be zero if mutual authentication is not supported.  Signature generation and verification is also used for Challenge and Measurements.

**Comment by steven-bellock** [comment]
@tmarkey64 
>The Validator will flag an error for 3.5.15 if the capabilities MUT_AUTH_CAP is not set and baseAsymAlg...

That `baseAsymAlg` is actually `ReqBaseAsymAlg`, ie the Requester's algorithm used to sign transcripts during mutual authentication. If mutual authentication is not supported by Requester then `ReqBaseAsymAlg` is `0` and that is reflected by the Responder.

**Comment by steven-bellock** [comment]
@tmarkey64 if this resolves your issue please close this issue.

**Comment by tmarkey64** [comment]
Thanks Steven, I have closed this issue.

**Comment by tmarkey64** [comment]
Hi Steven,
I examined the request for 3.5.15 (spdm_test_case_algorithms_success_11), the 3rd Algorithm Request Structure has AlgType set to 4 (ReqBaseAsymAlg), the AlgSupported field is set to 0x1FF.  Here is the raw request:

00000000 11 E3 04 00 30 00 01 00 FF 01 00 00 3F 00 00 00
00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00000020 02 20 3F 00 03 20 07 00 04 20 FF 01 05 20 01 00
00000030 45 76 65 00 00 00 00 00 00 00 00 00 00 00 00 00 

The responder in this case selected from the AlgSupported field, but the responder's capability flags did not include the MUT_AUTH_CAP, so it fails here (spdm_responder_test3_algorithms.c), where req_base_asym_alg is set to 0x80, which is included in the requesters bit map of supported assymetric algorithms above.

`    if (((test_buffer->rsp_cap_flags & SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_MUT_AUTH_CAP) != 0) &&
        ((req_base_asym_alg != 0xFFFF) && (req_base_asym_alg != 0x0))) {
        test_result = COMMON_TEST_RESULT_PASS;
    } else if (((test_buffer->rsp_cap_flags & SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_MUT_AUTH_CAP) ==
                0) &&
               (req_base_asym_alg == 0x0)) {
        test_result = COMMON_TEST_RESULT_PASS;
    } else {
        test_result = COMMON_TEST_RESULT_FAIL;  << -- fails here.
    }`

**Comment by steven-bellock** [comment]
>The responder in this case selected from the AlgSupported field, but the responder's capability flags did not include the MUT_AUTH_CAP, so it fails here

That is correct. If Responder's `MUT_AUTH_CAP` is `0` then Responder's `ReqBaseAsymAlg` should be `0` or that field should not exist in the response.

**Comment by tmarkey64** [comment]
Hi Steven, Thanks for the clarification.  I see now that this clarification was added in the 1.2 version of the specification.  "If the Responder is generating the signature, the selected cryptographic signing algorithm is indicated in exactly one of BaseAsymSel or ExtAsymSel in ALGORITHMS message. If the Requester is generating the signature, the selected cryptographic signing algorithm is indicated in ReqBaseAsymAlg of RespAlgStruct in ALGORITHMS message.".  I will go ahead and close the issue again.

**Comment by tmarkey64** [comment]
Issue has clarified and resolved.

**Comment by chanss-park** [comment]
Hi Steven. Currently, Validator will process the following test scenario as fail.
- Requester sets `GET_CAPABILITIES` with `MUT_AUTH_CAP=0`.
- Responder sets `CAPABILITIES` with `MUT_AUTH_CAP=1`.
- Requester sends `NEGOTIATE_ALGORITHMS` with non-zero `ReqBaseAsymAlg`.
- Responder sends `ALGORITHMS` without `ReqBaseAsymAlg` field (because Requester did not set `MUT_AUTH_CAP`).

https://github.com/DMTF/SPDM-Responder-Validator/blob/626dfbd625453bf27a6fe3a3f8e50de344c675dc/library/spdm_responder_conformance_test_lib/spdm_responder_test_3_algorithms.c#L1380-L1389


I have two questions related to the test scenario above.
1. What does it mean that Requester sends non-zero `ReqBaseAsymAlg` when `MUT_AUTH_CAP=0`?
2. Should Responder process `ReqBaseAsymAlg` even though Requester did not set `MUT_AUTH_CAP`?
According to the following statement in SPDM 1.3, it seems that Responder can choose nothing for `ReqBaseAsymAlg`.
`If the Responder will not send any messages that require a signature, this value should be set to zero.`

**Comment by steven-bellock** [comment]
>- Requester sets GET_CAPABILITIES with MUT_AUTH_CAP=0.
>- Requester sends NEGOTIATE_ALGORITHMS with non-zero ReqBaseAsymAlg.

@jyao1 @chanss-park this is a bug in the validator. Requester's `ReqBaseAsymAlg` should not exist if Requester's `MUT_AUTH_CAP` is `0`. https://github.com/DMTF/SPDM-Responder-Validator/blob/626dfbd625453bf27a6fe3a3f8e50de344c675dc/library/spdm_responder_conformance_test_lib/spdm_responder_test_3_algorithms.c#L56-L65

**Comment by chanss-park** [comment]
> 2. Should Responder process `ReqBaseAsymAlg` even though Requester did not set `MUT_AUTH_CAP`?
>    According to the following statement in SPDM 1.3, it seems that Responder can choose nothing for `ReqBaseAsymAlg`.
>    `If the Responder will not send any messages that require a signature, this value should be set to zero.`

Hi Steven, thank you for checking.
For the 2nd question, how should Responder process the case ? Invalid request error or just ignore the `ReqBaseAsymAlg` ?

**Comment by jyao1** [comment]
> @jyao1 @chanss-park this is a bug in the validator. Requester's ReqBaseAsymAlg should not exist if Requester's MUT_AUTH_CAP is 0.

Yes. I agree it is a bug. If MUT_AUTH_CAP is 0, requester should set 0 to ReqBaseAsymAlg.

**Comment by steven-bellock** [comment]
>For the 2nd question, how should Responder process the case ? Invalid request error or just ignore the ReqBaseAsymAlg ?

That is apparently a complicated question. Let me review the specification(s), as `ReqBaseAsymAlg` has had special treatment throughout the years.

**Comment by steven-bellock** [comment]
>Yes. I agree it is a bug. If MUT_AUTH_CAP is 0, requester should set 0 to ReqBaseAsymAlg, or make it absent.

Requester cannot set it `0`. The `AlgType` must be absent.
>For the 2nd question, how should Responder process the case ? Invalid request error or just ignore the ReqBaseAsymAlg ?

Yes, that should be an error.  Interestingly libspdm does not handle it that way precisely. I will file an issue against libspdm.

**Comment by chanss-park** [comment]
Hi Steven,
thank you for clarification.

**Comment by tmarkey64** [comment]
Hi Steven,

Regarding the ReqBaseAsymAlgo, the SPDM Specification states that, "If the
capabilities do not support
this algorithm, this value is not used and shall be set to zero."  The
specification does not state what capabilities flags are required to be set
in order to set ReqBaseAsym algorithm.  Both Challenge and Measurements
include a signature generation using an Asymmetric signing algorithm, so, I
am not sure that MUT_AUTH is the only capability that indicates that
ReqBaseAsym can be non-zero.  For example, if CHAL_CAP is set, wouldn't
this also indicate that this algorithm is supported?

**Comment by steven-bellock** [comment]
>Both Challenge and Measurements

Requester does not support `GET_MEASUREMENTS` request from the Responder. The three capabilities that determine if a Requester signs a message are
- `CHAL_CAP` (SPDM 1.1)
- `KEY_EX_CAP` (SPDM 1.1)
- `EP_INFO_CAP == 10b` (SPDM 1.3)

If a Requester sets `CHAL_CAP` and or `KEY_EX_CAP` then `MUT_AUTH_CAP` must also be set. `EP_INFO_CAP` does not require `MUT_AUTH_CAP`.

**Comment by tmarkey64** [comment]
Hi Steven,

Thank you for your input regarding the Requester needing to support
MUT_AUTH_CAP. I understand your point, but I believe the issue lies
elsewhere. The BaseAsymAlgo is defined for asymmetric key signature
algorithms used in signature verification. This is essential for both
mutual authentication and verifying Challenge Auth signatures.  The
Responder can support CHAL_CAP without supporting MUT_AUTH_CAP.

For this implementation scenario, Validator tests 3.5.15 and 3.6.15 will
fail as the test Requester does not set MUT_AUTH_CAP, and sets BaseAsymAlgo
to 0xFFF and Responder returns a non-zero BaseAsymSel.  The test then fails
as it requires ReqBaseAsym must be zero if MUT_AUTH_CAP is zero.  I do not
believe this is a valid failure as the Responder should be selecting the
BaseAsymSel from the supported BaseAsymAlgo for CHAL_CAP support.

Thanks, Tim Markey

**Comment by steven-bellock** [comment]
>The Responder can support CHAL_CAP without supporting MUT_AUTH_CAP.

Yes. The same is not true for the Requester.

>Validator tests 3.5.15 and 3.6.15 will fail as the test Requester does not set MUT_AUTH_CAP...

It should always set it in https://github.com/DMTF/SPDM-Responder-Validator/blob/31838b205b3bf24a0bcfa6aa45d7f92f080f4744/library/spdm_responder_conformance_test_lib/spdm_responder_test_3_algorithms.c#L60

@tmarkey64 feel free to join the meeting tomorrow morning. It might be easier to discuss this there.

**Comment by tmarkey64** [comment]
Hi Steven,

I am running the Validator with the test update which does set Requester
MUT_AUTH_CAP, however, since the Responder is not setting the MUT_AUTH_CAP,
the test still fails here (line 1381):

    if (((test_buffer->rsp_cap_flags &
SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_MUT_AUTH_CAP) != 0) &&
        ((req_base_asym_alg != 0xFFFF) && (req_base_asym_alg != 0x0))) {
        test_result = COMMON_TEST_RESULT_PASS;
    } else if (((test_buffer->rsp_cap_flags &
SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_MUT_AUTH_CAP) ==
                0) &&
               (req_base_asym_alg == 0x0)) {
        test_result = COMMON_TEST_RESULT_PASS;
    } else {
        test_result = COMMON_TEST_RESULT_FAIL;
    }

In this case,

test_buffer->rsp_cap_flags = 0x16 (CHAL_CAP, MEAS_CAP. CERT_CAP)
req_base_asym_alg = 0x80

I am not sure how to join the meeting, but I will look into this.
Thanks,
Tim

**Comment by steven-bellock** [comment]
>req_base_asym_alg = 0x80

The Responder is setting this value? That is not correct as the Responder does not have a capability to validate a Requester's signed message. It needs `MUT_AUTH_CAP` for that, at least for SPDM <= 1.2.

**Comment by tmarkey64** [comment]
Hi Steven,

The req_base_asym_alg variable is in SPDM Validator test function
spdm_test_case_algorithms_success_11.  This variable is set by the function
from the
Algorithms response variable struct_table_alg_supported.  The test
Requester sends the following Negotiate Algorithms request:

00000000 11 E3 04 00 30 00 01 00 FF 01 00 00 3F 00 00 00
00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00000020 02 20 3F 00 03 20 07 00 04 20 FF 01 05 20 01 00
00000040 43 6F 6D 00 00 00 00 00 00 00 00 00 00 00 00 00

Since the Requester has sent that it supports ReqBaseAsymAlg (offset 0x28),
the Responder returns the ReqBaseAsymAlg with AlgSupported set to 0x80.
Since, the bitmask in the Responder's Algorithms response is for "Bit mask
for indicating a Requester-supported, Responder-selected, SPDM-enumerated
asymmetric key signature algorithm for the purpose of signature
verification.", and the Responder supports Challenge_Auth, I believe this
response is correct.  The specification does not state whether this
response for signature verification is verification of a signature by the
Responder, I believe it can also be used to communicate to the Requester
the signature verification algorithm to use to verify the Challenge Auth
signature.

Thanks,
Tim

**Comment by steven-bellock** [comment]
>The specification does not state whether this response for signature verification is verification of a signature by the Responder,

I think I brought this up a very long time ago, but it sounds like it should be clarified in the specification:
- `BaseAsymAlgo / BaseAsymSel` - algorithm used by the Responder to sign messages, and used by the Requester to verify signatures.
- `ReqBaseAsymAlg` - algorithm used by the Requester to sign messages, and used by the Responder to verify signatures.

I'll file an issue against the specification.

**Comment by steven-bellock** [comment]
In particular `BaseAsymAlgo` would go from
>Shall be the bit mask listing Requester-supported SPDM-enumerated asymmetric key signature algorithms for the purpose of signature verification.

to (emphasis mine)
>Shall be the bit mask listing Requester-supported SPDM-enumerated asymmetric key signature algorithms for the purpose of signature verification **by the Requester**.

**Comment by tmarkey64** [comment]
Thanks Steven for the clarification.  Would this change expected to be
retro-active to older SPDM versions?
Tim

**Comment by steven-bellock** [comment]
Yes, it would go all the way back to 1.0.
